src/host/getset.cpp

```diff
@@ -913,53 +987,82 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
+            {
+                WriteClearScreen(context);
+                writer.Submit();
+                return S_OK;
+            }
 
-        // GH#3126 - This is a shim for cmd's `cls` function. In the
-        // legacy console, `cls` is supposed to clear the entire buffer. In
-        // conpty however, there's no difference between the viewport and the
-        // entirety of the buffer. We're going to see if this API call exactly
-        // matched the way we expect cmd to call it. If it does, then
-        // let's manually emit a ^[[3J to the connected terminal, so that their
-        // entire buffer will be cleared as well.
-        auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-        if (enableCmdShim && gci.IsInVtIoMode())
-        {
-            const auto currentBufferDimensions = buffer.GetBufferSize().Dimensions();
-            const auto sourceIsWholeBuffer = (source.top == 0) &&
-                                             (source.left == 0) &&
-                                             (source.right == currentBufferDimensions.width) &&
-                                             (source.bottom == currentBufferDimensions.height);
-            const auto targetIsNegativeBufferHeight = (target.x == 0) &&
-                                                      (target.y == -currentBufferDimensions.height);
-            const auto noClipProvided = clip == std::nullopt;
-            const auto fillIsBlank = (fillCharacter == UNICODE_SPACE) &&
-                                     (fillAttribute == buffer.GetAttributes().GetLegacyAttributes());
-
-            if (sourceIsWholeBuffer && targetIsNegativeBufferHeight && noClipProvided && fillIsBlank)
+            const auto clipViewport = clip ? Viewport::FromInclusive(*clip) : bufferSize;
+            const auto sourceViewport = Viewport::FromInclusive(source);
+            Viewport readViewport;
+            Viewport writtenViewport;
+
+            const auto w = std::max(0, sourceViewport.Width());
+            const auto h = std::max(0, sourceViewport.Height());
+            const auto a = static_cast<size_t>(w * h);
+            if (a == 0)
             {
-                // It's important that we flush the renderer at this point so we don't
-                // have any pending output rendered after the scrollback is cleared.
-                ServiceLocator::LocateGlobals().pRender->TriggerFlush(false);
-                hr = gci.GetVtIo()->ManuallyClearScrollback();
+                return S_OK;
             }
+
+            til::small_vector<CHAR_INFO, 1024> backup;
+            til::small_vector<CHAR_INFO, 1024> fill;
+
+            backup.resize(a, CHAR_INFO{ fillCharacter, fillAttribute });
+            fill.resize(a, CHAR_INFO{ fillCharacter, fillAttribute });
+
+            writer.BackupCursor();
+
+            RETURN_IF_FAILED(ReadConsoleOutputWImplHelper(context, backup, sourceViewport, readViewport));

```

Comment:
I'm a little skeptical that such a structure could be defined, given the breadth of the VT API landscape. For instance, scrolling a sixel image should work (it works with the VT APIs), so how do you express this in terms of a `CHAR_INFO_EX` structure?

This makes me believe that the idea of lossless scrolling of buffer contents with the console APIs is long-term not viable, unless the request can be cleanly translated to VT sequences. It's possible that we may be able to do that in the future (for instance with SU/SD as mentioned by Dustin yesterday.)

Reply:
`ScrollConsoleScreenBuffer` is really more of a copy operation plus fill, rather than a scroll, and I think should be cleanly translatable to `DECCRA` plus `DECFRA`.